File: src/main/paradox/lagom/code/jdocs/lagom/FormingACluster.java
package jdocs.lagom;

import com.google.inject.AbstractModule;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.api.Service;
import com.lightbend.lagom.javadsl.server.ServiceGuiceSupport;

public interface FormingACluster {

    interface start {

        public class ShoppingCartModule extends AbstractModule implements ServiceGuiceSupport {
            //#start
            @Override
            protected void configure() {
                bindService(ShoppingCartService.class, ShoppingCartServiceImpl.class);
                bind(ClusterBootstrapStart.class).asEagerSingleton();
            }
            //#start
        }

        interface ShoppingCartService extends Service {
            @Override
            default Descriptor descriptor() {
                return Service.named("shopping-cart");
            }
        }

        class ShoppingCartServiceImpl implements ShoppingCartService {}

    }
}

File: src/main/paradox/lagom/code/FormingACluster.scala
package sdocs.lagom

import com.lightbend.lagom.scaladsl.api.{Descriptor, Service}
import com.lightbend.lagom.scaladsl.devmode.LagomDevModeComponents
import com.lightbend.lagom.scaladsl.server.{LagomApplication, LagomApplicationContext, LagomApplicationLoader, LagomServer}
import play.api.libs.ws.ahc.AhcWSComponents

object StartClusterBootstrap {

  class ShoppingCartLoader extends LagomApplicationLoader {

    //#start
    import akka.management.scaladsl.AkkaManagement
    import akka.management.cluster.bootstrap.ClusterBootstrap
    import com.lightbend.lagom.scaladsl.akka.discovery.AkkaDiscoveryComponents

    override def load(context: LagomApplicationContext): LagomApplication =
      new ShoppingCartApplication(context) with AkkaDiscoveryComponents {
        AkkaManagement(actorSystem).start()
        ClusterBootstrap(actorSystem).start()
      }
    //#start

    override def loadDevMode(context: LagomApplicationContext): LagomApplication =
      new ShoppingCartApplication(context) with LagomDevModeComponents

    override def describeService = Some(readDescriptor[ShoppingCartService])
  }

  abstract class ShoppingCartApplication(context: LagomApplicationContext)
    extends LagomApplication(context) with AhcWSComponents {
    override lazy val lagomServer: LagomServer = serverFor[ShoppingCartService](new ShoppingCartService {})
  }

  trait ShoppingCartService extends Service {
    override def descriptor: Descriptor = Service.named("shopping-cart")
  }

}
